xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc

```diff
@@ -0,0 +1,1708 @@
+// Copyright 2024 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.h"
+
+#include <cassert>
+#include <cstdint>
+
+#include "absl/status/status.h"
+#include "absl/strings/str_cat.h"
+#include "llvm/include/llvm/Support/SourceMgr.h"
+#include "llvm/include/llvm/Support/raw_ostream.h"
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/include/mlir/IR/BuiltinOps.h"
+#include "mlir/include/mlir/IR/Location.h"
+#include "xls/contrib/mlir/IR/xls_ops.h"
+#include "xls/ir/function.h"
+#include "xls/ir/function_base.h"
+#include "xls/ir/nodes.h"
+#include "xls/ir/proc.h"
+#include "xls/public/ir_parser.h"
+
+namespace mlir::xls {
+
+//===----------------------------------------------------------------------===//
+// Translation State
+//===----------------------------------------------------------------------===//
+
+class TranslationState {
+ public:
+  TranslationState() = default;
+
+  absl::Status setFileName(::xls::Fileno num, StringAttr name) {
+    if (fileno_map_.contains(num)) {
+      return absl::InternalError(
+          absl::StrCat("Duplicate file number ", num.value()));
+    }
+    fileno_map_[num] = name;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<StringAttr> getFileName(::xls::Fileno no) {
+    auto op = fileno_map_.find(no);
+    if (op == fileno_map_.end()) {
+      return absl::InternalError(
+          absl::StrCat("Unknown file number  ", no.value()));
+    } else {
+      return op->second;
+    }
+  }
+
+  absl::Status setFn(std::string name, FlatSymbolRefAttr fn) {
+    if (fn_map_.contains(name)) {
+      return absl::InternalError(
+          absl::StrCat("Duplicate function name ", name));
+    }
+    fn_map_[name] = fn;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<FlatSymbolRefAttr> getFn(std::string name) {
+    auto op = fn_map_.find(name);
+    if (op == fn_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown function name  ", name));
+    } else {
+      return op->second;
+    }
+  }
+
+  absl::Status setChn(std::string name, FlatSymbolRefAttr chn) {
+    if (chn_map_.contains(name)) {
+      return absl::InternalError(absl::StrCat("Duplicate channel name ", name));
+    }
+    chn_map_[name] = chn;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<SymbolRefAttr> getChn(std::string name) {
```

Comment:
Please call this getChannel ; avoid unnecessary abbreviations as they hinder readability.